segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/ContainerKeyIndex.java

```diff
@@ -664,7 +667,7 @@ private void triggerCacheTailIndex(DirectSegmentAccess segment, long lastIndexed
                 .thenApplyAsync(inputData -> {
                     // Parse out all Table Keys and collect their latest offsets, as well as whether they were deleted.
                     val updates = new TailUpdates();
-                    collectLatestOffsets(inputData, startOffset, maxLength, updates);
+                    collectLatestOffsetsWithHighestVersion(inputData, startOffset, maxLength, updates, tailCachePreIndexVersionTracker);
 
                     // Incorporate that into the cache.
                     this.cache.includeTailCache(segment.getSegmentId(), updates.byBucket);
```

Comment:
If I understand the description of the PR correctly, the map is to guard this method so that we don't include old elements into the cache overriding newer ones inside this call.

If this is correct, can we do the check in here? It looks like this call is specific to the tail cache. This would allow the `updates` to server the same purpose as `tailCachePreIndexVersionTracker`, and the invariant could be enforced explicitly rather than relying on the caller.

Reply:
You are right, the problem is just related to the tail-caching process. Regular PUTs that update the tail-cache cannot fall into this situation, because they are unaware of compacted entries. However, during the recovery of the Table Segment, we scan the segment and have to deal with both regular PUTs and compacted entries. Failing to enforce the invariants in this specific recovery process was the reason for serving stale data (at least for some time) on GETs. In the rest of the regular PUT/GET path, things appear to be done correctly.